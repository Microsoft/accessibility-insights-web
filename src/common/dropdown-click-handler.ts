// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { DropdownActionMessageCreator } from './message-creators/dropdown-action-message-creator';
import { TelemetryEventSource } from './types/telemetry-data';

export class DropdownClickHandler {
    private dropdownActionMessageCreator: DropdownActionMessageCreator;
    private source: TelemetryEventSource;

    constructor(
        dropdownActionMessageCreator: DropdownActionMessageCreator,
        source: TelemetryEventSource,
    ) {
        this.source = source;
        this.dropdownActionMessageCreator = dropdownActionMessageCreator;
    }

    public openPreviewFeaturesPanelHandler = (event: React.MouseEvent<HTMLElement>): void => {
        this.dropdownActionMessageCreator.openPreviewFeaturesPanel(event, this.source);
    };

    public openScopingPanelHandler = (event: React.MouseEvent<HTMLElement>): void => {
        this.dropdownActionMessageCreator.openScopingPanel(event, this.source);
    };

    public openSettingsPanelHandler = (event: React.MouseEvent<HTMLElement>): void => {
        this.dropdownActionMessageCreator.openSettingsPanel(event, this.source);
    };

    public openDebugTools = (): void => {
        this.dropdownActionMessageCreator.openDebugTools();
    };
}
