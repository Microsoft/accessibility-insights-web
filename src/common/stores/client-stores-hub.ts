// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { IBaseStore } from '../istore';

export interface ClientStoresHub<T> {
    stores: IBaseStore<any>[];
    addChangedListenerToAllStores(listener: () => void): void;
    removeChangedListenerFromAllStores(listener: () => void): void;
    hasStores(): boolean;
    hasStoreData(): boolean;
    getAllStoreData(): T;
}
