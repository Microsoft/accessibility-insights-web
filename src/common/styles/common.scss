// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import 'high-contrast-override.scss';
@import 'colors.scss';
@import 'fonts.scss';

$focus-ring-color: $neutral-100;

@mixin set-focus-ring-styles($property: 'border') {
    #{$property}: 1px solid $focus-ring-color;
    @include ms-high-contrast-override {
        #{$property}: 1px dashed windowtext !important;
    }
}

.ms-Fabric.is-focusVisible .ms-Nav-group .ms-nav-linkButton:focus::after {
    @include set-focus-ring-styles('border');
}

button::after {
    content: '';
    position: absolute;
}

.ms-Button.ms-Button--primary:not(.is-disabled) {
    background: $communication-primary;

    &:hover {
        background: $communication-shade-10;
    }
}

.ms-fontColor-neutralPrimary,
.ms-fontColor-neutralPrimary--hover:hover {
    color: $neutral-100 !important;
}

.ms-Fabric {
    color: $neutral-100;
}

.generic-toggle-component {
    margin-bottom: 4vh;
    .toggle-container {
        margin-bottom: 1vh;
        .toggle-name {
            display: inline-block;
            font-size: 17px;
            font-weight: bold;
        }

        .toggle {
            float: right;
        }
    }
}

.high-contrast-theme {
    .insights-link {
        &:hover {
            text-decoration: underline;
        }
    }
    .is-selected {
        .status-icon {
            color: $high-contrast-icon-color !important;
        }
    }
}

.insights-link {
    // marking important to override office fabric style
    color: $link !important;
    text-decoration: none;

    &:hover {
        color: $link-hover !important;
        text-decoration: underline;
    }
}

.ms-Spinner {
    .ms-Spinner-circle {
        border-top-color: $spinner-text;
    }

    .ms-Spinner-label {
        color: $spinner-text;
    }
}

$normalTitleFontWeight: 300 !default;

$pivotItemLeftBorderWidth: 3px !default;
$pivotItemBorderStyle: solid !default;

$detailsViewHeaderBarHeight: 40px !default;
$detailsViewNavPivotsHeight: 41px !default;
$detailsViewCommandBarHeight: 40px !default;
$detailsViewCommandBarBorderHeight: 1px !default;
$detailsViewLeftNavWidth: 200px !default;
$detailsViewTotalHeaderHeight: (#{$detailsViewCommandBarBorderHeight} + #{$detailsViewCommandBarHeight} + #{$detailsViewHeaderBarHeight});

$fastPassRightPanelMarginRight: 24px !default;
$fastPassRightPanelMarginLeft: 24px !default;
$fastPassRightPanelMarginTop: 24px !default;

@mixin ellipsedText {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}
