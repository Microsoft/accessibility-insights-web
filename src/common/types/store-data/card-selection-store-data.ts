// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export interface RuleExpandCollapseData {
    isExpanded: boolean;
    cards: { [resultInstanceUid: string]: boolean };
}

export interface CardSelectionStoreData {
    rules: { [ruleId: string]: RuleExpandCollapseData };
}
