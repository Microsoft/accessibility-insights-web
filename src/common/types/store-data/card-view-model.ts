// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { GuidanceLink } from '../../../scanner/rule-to-links-mappings';
import { InstanceResultStatus, UnifiedResult } from './unified-data-interface';
export type CardRuleResultStatus = InstanceResultStatus | 'inapplicable';
export interface CardRuleResult {
    id: string;
    nodes: CardResult[];
    description: string;
    url: string;
    guidance: GuidanceLink[];
    isExpanded: boolean;
}
export type CardRuleResultsByStatus = {
    [key in CardRuleResultStatus]: CardRuleResult[];
};
export interface CardResult extends UnifiedResult {
    isSelected: boolean;
}
export const AllRuleResultStatuses: CardRuleResultStatus[] = ['pass', 'fail', 'unknown', 'inapplicable'];
