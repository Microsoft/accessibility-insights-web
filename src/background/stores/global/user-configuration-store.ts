// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { autobind } from '@uifabric/utilities';
import { cloneDeep } from 'lodash';

import { FeatureFlags } from '../../../common/feature-flags';
import { IndexedDBAPI } from '../../../common/indexedDB/indexedDB';
import { StoreNames } from '../../../common/stores/store-names';
import { UserConfigurationStoreData } from '../../../common/types/store-data/user-configuration-store';
import {
    SetBugServicePayload,
    SetBugServicePropertyPayload,
    SetHighContrastModePayload,
    SetIssueTrackerPathPayload,
    SetTelemetryStatePayload,
} from '../../actions/action-payloads';
import { FeatureFlagPayload } from '../../actions/feature-flag-actions';
import { UserConfigurationActions } from '../../actions/user-configuration-actions';
import { IndexedDBDataKeys } from '../../IndexedDBDataKeys';
import { BaseStore } from '../base-store';

export class UserConfigurationStore extends BaseStore<UserConfigurationStoreData> {
    public static readonly defaultState: UserConfigurationStoreData = {
        isFirstTime: true,
        enableTelemetry: false,
        enableHighContrast: false,
        bugService: 'none',
        bugServicePropertiesMap: {},
    };

    constructor(
        private readonly initialState: UserConfigurationStoreData,
        private readonly userConfigActions: UserConfigurationActions,
        private readonly indexDbApi: IndexedDBAPI,
    ) {
        super(StoreNames.UserConfigurationStore);
    }

    public getDefaultState(): UserConfigurationStoreData {
        return this.initialState ? cloneDeep(this.initialState) : cloneDeep(UserConfigurationStore.defaultState);
    }

    protected addActionListeners(): void {
        this.userConfigActions.getCurrentState.addListener(this.onGetCurrentState);
        this.userConfigActions.setTelemetryState.addListener(this.onSetTelemetryState);
        this.userConfigActions.setHighContrastMode.addListener(this.onSetHighContrastMode);
        this.userConfigActions.setBugService.addListener(this.onSetBugService);
        this.userConfigActions.setBugServiceProperty.addListener(this.onSetBugServiceProperty);
        this.userConfigActions.notifyFeatureFlagChange.addListener(this.onNotifyFeatureFlagChange);
        this.userConfigActions.setIssueTrackerPath.addListener(this.onSetIssueTrackerPath);
    }

    @autobind
    private onSetTelemetryState(payload: SetTelemetryStatePayload): void {
        this.state.isFirstTime = false;
        this.state.enableTelemetry = payload.enableTelemetry;

        // tslint:disable-next-line:no-floating-promises - grandfathered-in pre-existing violation
        this.indexDbApi.setItem(IndexedDBDataKeys.userConfiguration, this.state);
        this.emitChanged();
    }

    @autobind
    private onSetHighContrastMode(payload: SetHighContrastModePayload): void {
        this.state.enableHighContrast = payload.enableHighContrast;

        // tslint:disable-next-line:no-floating-promises - grandfathered-in pre-existing violation
        this.indexDbApi.setItem(IndexedDBDataKeys.userConfiguration, this.state);
        this.emitChanged();
    }

    @autobind
    private onSetBugService(payload: SetBugServicePayload): void {
        this.state.bugService = payload.bugServiceName;

        // tslint:disable-next-line:no-floating-promises - grandfathered-in pre-existing violation
        this.indexDbApi.setItem(IndexedDBDataKeys.userConfiguration, this.state);
        this.emitChanged();
    }

    @autobind
    private onSetBugServiceProperty(payload: SetBugServicePropertyPayload): void {
        if (!this.state.bugServicePropertiesMap) {
            this.state.bugServicePropertiesMap = {};
        }
        if (!this.state.bugServicePropertiesMap[payload.bugServiceName]) {
            this.state.bugServicePropertiesMap[payload.bugServiceName] = {};
        }

        this.state.bugServicePropertiesMap[payload.bugServiceName][payload.propertyName] = payload.propertyValue;

        // tslint:disable-next-line:no-floating-promises - grandfathered-in pre-existing violation
        this.indexDbApi.setItem(IndexedDBDataKeys.userConfiguration, this.state);
        this.emitChanged();
    }

    @autobind
    private onNotifyFeatureFlagChange(payload: FeatureFlagPayload): void {
        if (payload.feature === FeatureFlags.highContrastMode && payload.enabled === false) {
            this.onSetHighContrastMode({ enableHighContrast: false });
        }
    }

    @autobind
    private onSetIssueTrackerPath(payload: SetIssueTrackerPathPayload) {
        this.state.issueTrackerPath = payload.issueTrackerPath;

        // tslint:disable-next-line:no-floating-promises - grandfathered-in pre-existing violation
        this.indexDbApi.setItem(IndexedDBDataKeys.userConfiguration, this.state);
        this.emitChanged();
    }
}
