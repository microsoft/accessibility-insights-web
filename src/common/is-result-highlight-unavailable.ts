// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { FilterableResult } from 'common/get-card-selection-view-data';
import { PlatformData } from 'common/types/store-data/unified-data-interface';

export type IsResultHighlightUnavailable = (
    result: FilterableResult,
    platformInfo: PlatformData | null,
) => boolean;

export const isResultHighlightUnavailableWeb: IsResultHighlightUnavailable = () => false;
