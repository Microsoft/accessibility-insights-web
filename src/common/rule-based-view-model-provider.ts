// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { UnifiedRuleResult, UnifiedStatusResults } from './components/cards/failed-instances-section';
import { AllRuleResultStatuses, UnifiedResult, UnifiedRule, UnifiedRuleResultStatus } from './types/store-data/unified-data-interface';

export type GetUnifiedRuleResultsDelegate = (rules: UnifiedRule[], results: UnifiedResult[]) => UnifiedStatusResults;

export const getUnifiedRuleResults: GetUnifiedRuleResultsDelegate = function(
    rules: UnifiedRule[],
    results: UnifiedResult[],
): UnifiedStatusResults {
    if (results == null || rules == null) {
        return null;
    }

    const statusResults = getEmptyStatusResults();
    const ruleIdsWithResultNodes: Set<string> = new Set();

    for (const result of results) {
        const ruleResults = statusResults[result.status];
        const ruleResultIndex: number = getRuleResultIndex(result, ruleResults);

        if (ruleResultIndex !== -1) {
            ruleResults[ruleResultIndex].nodes.push(result);
        } else {
            const unifiedRule: UnifiedRule = getUnifiedRule(result.ruleId, rules);
            if (unifiedRule) {
                ruleResults.push(createUnifiedRuleResult(result, unifiedRule));
            }
        }

        ruleIdsWithResultNodes.add(result.ruleId);
    }

    for (const rule of rules) {
        if (!ruleIdsWithResultNodes.has(rule.id)) {
            statusResults.inapplicable.push(createRuleResultWithoutNodes('inapplicable', rule));
        }
    }

    return statusResults;
};

function getEmptyStatusResults(): UnifiedStatusResults {
    const statusResults = {};

    AllRuleResultStatuses.forEach(status => {
        statusResults[status] = [];
    });

    return statusResults as UnifiedStatusResults;
}

function createUnifiedRuleResult(result: UnifiedResult, rule: UnifiedRule): UnifiedRuleResult {
    return {
        id: rule.id,
        status: result.status,
        nodes: [result],
        description: rule.description,
        url: rule.url,
        guidance: rule.guidance,
    };
}

function createRuleResultWithoutNodes(status: UnifiedRuleResultStatus, rule: UnifiedRule): UnifiedRuleResult {
    return {
        id: rule.id,
        status: status,
        nodes: [],
        description: rule.description,
        url: rule.url,
        guidance: rule.guidance,
    };
}

function getUnifiedRule(id: string, rules: UnifiedRule[]): UnifiedRule {
    return rules.find(rule => rule.id === id);
}

function getRuleResultIndex(result: UnifiedResult, ruleResults: UnifiedRuleResult[]): number {
    return ruleResults.findIndex(ruleResult => ruleResult.id === result.ruleId);
}
