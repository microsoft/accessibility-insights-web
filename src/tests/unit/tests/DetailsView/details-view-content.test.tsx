// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DropdownClickHandler } from 'common/dropdown-click-handler';
import {
    CardSelectionViewData,
    GetCardSelectionViewData,
} from 'common/get-card-selection-view-data';
import { IsResultHighlightUnavailable } from 'common/is-result-highlight-unavailable';
import { GetCardViewData } from 'common/rule-based-view-model-provider';
import { ClientStoresHub } from 'common/stores/client-stores-hub';
import { CardSelectionStoreData } from 'common/types/store-data/card-selection-store-data';
import { CardsViewModel } from 'common/types/store-data/card-view-model';
import { DetailsViewPivotType } from 'common/types/store-data/details-view-pivot-type';
import { DetailsViewRightContentPanelType } from 'common/types/store-data/details-view-right-content-panel-type';
import {
    TargetAppData,
    ToolData,
    UnifiedScanResultStoreData,
} from 'common/types/store-data/unified-data-interface';
import { UserConfigurationStoreData } from 'common/types/store-data/user-configuration-store';
import { VisualizationType } from 'common/types/visualization-type';
import { DetailsViewActionMessageCreator } from 'DetailsView/actions/details-view-action-message-creator';
import { DetailsViewContent } from 'DetailsView/components/details-view-content';
import {
    DetailsRightPanelConfiguration,
    GetDetailsRightPanelConfiguration,
    GetDetailsRightPanelConfigurationProps,
} from 'DetailsView/components/details-view-right-panel';
import {
    DetailsViewSwitcherNavConfiguration,
    GetDetailsSwitcherNavConfiguration,
    GetDetailsSwitcherNavConfigurationProps,
} from 'DetailsView/components/details-view-switcher-nav';
import { GetSelectedAssessmentStoreData } from 'DetailsView/components/left-nav/get-selected-assessment-store-data';
import { GetSelectedDetailsViewProps } from 'DetailsView/components/left-nav/get-selected-details-view';
import {
    DetailsViewContainerDeps,
    DetailsViewContainerState,
} from 'DetailsView/details-view-container';
import { AssessmentInstanceTableHandler } from 'DetailsView/handlers/assessment-instance-table-handler';
import { DetailsViewToggleClickHandlerFactory } from 'DetailsView/handlers/details-view-toggle-click-handler-factory';
import { shallow } from 'enzyme';
import * as React from 'react';
import { IMock, It, Mock, MockBehavior, Times } from 'typemoq';

import { DetailsViewStoreDataBuilder } from '../../common/details-view-store-data-builder';
import { VisualizationStoreDataBuilder } from '../../common/visualization-store-data-builder';
import { DetailsViewContainerPropsBuilder } from './details-view-container-props-builder';
import { StoreMocks } from './store-mocks';

describe(DetailsViewContent.displayName, () => {
    const pageTitle = 'DetailsViewContainerTest title';
    const pageUrl = 'http://detailsViewContainerTest/url/';
    let detailsViewActionMessageCreator: IMock<DetailsViewActionMessageCreator>;
    let deps: DetailsViewContainerDeps;
    let getDetailsRightPanelConfiguration: IMock<GetDetailsRightPanelConfiguration>;
    let getDetailsSwitcherNavConfiguration: IMock<GetDetailsSwitcherNavConfiguration>;
    let getCardViewDataMock: IMock<GetCardViewData>;
    let getCardSelectionViewDataMock: IMock<GetCardSelectionViewData>;
    let targetAppInfo: TargetAppData;
    let isResultHighlightUnavailableStub: IsResultHighlightUnavailable;
    let timestamp: string;
    let scanDate: Date;
    let toolData: ToolData;
    let getDateFromTimestampMock: IMock<(timestamp: string) => Date>;

    beforeEach(() => {
        detailsViewActionMessageCreator = Mock.ofType(DetailsViewActionMessageCreator);
        getDetailsRightPanelConfiguration = Mock.ofInstance(
            (props: GetDetailsRightPanelConfigurationProps) => null,
            MockBehavior.Strict,
        );
        getDetailsSwitcherNavConfiguration = Mock.ofInstance(
            (props: GetDetailsSwitcherNavConfigurationProps) => null,
            MockBehavior.Strict,
        );
        getCardViewDataMock = Mock.ofInstance(
            (
                storeData: UnifiedScanResultStoreData,
                cardSelectionViewData?: CardSelectionViewData,
            ) => null,
            MockBehavior.Strict,
        );
        getCardSelectionViewDataMock = Mock.ofInstance(
            (storeData: CardSelectionStoreData) => null,
            MockBehavior.Strict,
        );
        isResultHighlightUnavailableStub = () => null;
        timestamp = 'timestamp';
        scanDate = new Date(Date.UTC(0, 1, 2, 3));
        getDateFromTimestampMock = Mock.ofInstance(() => null);
        getDateFromTimestampMock.setup(gd => gd(timestamp)).returns(() => scanDate);
        targetAppInfo = {
            name: pageTitle,
            url: pageUrl,
        };
        toolData = {
            applicationProperties: { name: 'some app' },
        } as ToolData;

        const assessmentInstanceTableHandlerMock = Mock.ofType(AssessmentInstanceTableHandler);

        deps = {
            detailsViewActionMessageCreator: detailsViewActionMessageCreator.object,
            getDetailsRightPanelConfiguration: getDetailsRightPanelConfiguration.object,
            getDetailsSwitcherNavConfiguration: getDetailsSwitcherNavConfiguration.object,
            getCardViewData: getCardViewDataMock.object,
            getCardSelectionViewData: getCardSelectionViewDataMock.object,
            isResultHighlightUnavailable: isResultHighlightUnavailableStub,
            getDateFromTimestamp: getDateFromTimestampMock.object,
            getAssessmentInstanceTableHandler: () => assessmentInstanceTableHandlerMock.object,
        } as DetailsViewContainerDeps;
    });

    describe('render', () => {
        it('renders normally', () => {
            const viewType = -1;
            const isPreviewFeaturesOpen = false;
            const clickHandlerFactoryMock = Mock.ofType(DetailsViewToggleClickHandlerFactory);
            const dropdownClickHandler = Mock.ofType(DropdownClickHandler);
            const getSelectedDetailsViewMock = Mock.ofInstance(
                (theProps: GetSelectedDetailsViewProps) => null,
                MockBehavior.Strict,
            );
            const getSelectedAssessmentStoreDataMock = Mock.ofInstance(
                (() => null) as GetSelectedAssessmentStoreData,
            );

            const rightContentPanelType = 'TestView';
            const rightContentPanelConfig = {} as DetailsRightPanelConfiguration;
            const switcherNavConfig = {
                getSelectedDetailsView: getSelectedDetailsViewMock.object,
                getSelectedAssessmentStoreData: getSelectedAssessmentStoreDataMock.object,
            } as DetailsViewSwitcherNavConfiguration;

            const visualizationStoreData = new VisualizationStoreDataBuilder()
                .withSelectedDetailsViewPivot(DetailsViewPivotType.fastPass)
                .with('selectedAdhocDetailsView', viewType)
                .build();

            setupActionMessageCreatorMock(
                detailsViewActionMessageCreator,
                visualizationStoreData.selectedDetailsViewPivot,
                1,
            );

            setupGetDetailsRightPanelConfiguration(
                rightContentPanelType,
                visualizationStoreData.selectedDetailsViewPivot,
                rightContentPanelConfig,
            );

            setupGetSwitcherNavConfiguration(
                visualizationStoreData.selectedDetailsViewPivot,
                switcherNavConfig,
            );

            const detailsViewState = new DetailsViewStoreDataBuilder()
                .withPreviewFeaturesOpen(isPreviewFeaturesOpen)
                .withDetailsViewRightContentPanel(rightContentPanelType)
                .build();

            const userConfigurationStoreData: UserConfigurationStoreData = {
                isFirstTime: false,
                enableTelemetry: true,
                enableHighContrast: false,
                lastSelectedHighContrast: false,
                bugService: 'gitHub',
                bugServicePropertiesMap: { gitHub: { repository: 'gitHub-repository' } },
                adbLocation: null,
                lastWindowState: null,
                lastWindowBounds: null,
                showAutoDetectedFailuresDialog: true,
                showSaveAssessmentDialog: true,
            };

            const unifiedScanResultStoreData: UnifiedScanResultStoreData = {
                results: [],
                rules: [],
                targetAppInfo: targetAppInfo,
                timestamp: timestamp,
                toolInfo: toolData,
            };

            const storeMocks = new StoreMocks()
                .setVisualizationStoreData(visualizationStoreData)
                .setDetailsViewStoreData(detailsViewState)
                .setUserConfigurationStoreData(userConfigurationStoreData)
                .setUnifiedScanResultStoreData(unifiedScanResultStoreData)
                .setNeedsReviewScanResultStoreData(unifiedScanResultStoreData);

            const storesHubMock = createStoresHubMock(storeMocks);

            deps.dropdownClickHandler = dropdownClickHandler.object;

            const props = new DetailsViewContainerPropsBuilder(deps)
                .setStoreMocks(storeMocks)
                .setStoresHubMock(storesHubMock.object)
                .build();

            const state = getState(storeMocks, viewType, rightContentPanelConfig);

            getSelectedDetailsViewMock
                .setup(gsdvm =>
                    gsdvm(
                        It.isObjectWith({
                            assessmentStoreData: state.assessmentStoreData,
                            visualizationStoreData: state.visualizationStoreData,
                            quickAssessStoreData: state.quickAssessStoreData,
                        }),
                    ),
                )
                .returns(() => viewType);

            getSelectedAssessmentStoreDataMock
                .setup(m =>
                    m(
                        It.isObjectWith({
                            assessmentStoreData: state.assessmentStoreData,
                            quickAssessStoreData: state.quickAssessStoreData,
                        }),
                    ),
                )
                .returns(() => {
                    return state.assessmentStoreData;
                });

            const cardSelectionViewData: CardSelectionViewData = {} as CardSelectionViewData;
            getCardSelectionViewDataMock
                .setup(g =>
                    g(
                        state.cardSelectionStoreData,
                        state.unifiedScanResultStoreData.results,
                        state.unifiedScanResultStoreData.platformInfo,
                        isResultHighlightUnavailableStub,
                    ),
                )
                .returns(() => cardSelectionViewData)
                .verifiable(Times.exactly(2));

            const cardsViewData: CardsViewModel = {} as any;
            getCardViewDataMock
                .setup(m => m(state.unifiedScanResultStoreData, cardSelectionViewData))
                .returns(() => cardsViewData);

            const rendered = shallow(
                <DetailsViewContent
                    {...props}
                    narrowModeStatus={{
                        isHeaderAndNavCollapsed: false,
                        isCommandBarCollapsed: false,
                        isVirtualKeyboardCollapsed: false,
                        isCardFooterCollapsed: false,
                    }}
                    isSideNavOpen={false}
                    setSideNavOpen={() => {}}
                />,
            );
            expect(rendered.getElement()).toMatchSnapshot();

            clickHandlerFactoryMock.verifyAll();
            getCardSelectionViewDataMock.verifyAll();
        });
    });

    function createStoresHubMock(
        storeMocks: StoreMocks,
        hasStores = true,
        hasStoreData = true,
    ): IMock<ClientStoresHub<any>> {
        const storesHubMock = Mock.ofType(ClientStoresHub);
        storesHubMock.setup(s => s.hasStores()).returns(() => hasStores);
        storesHubMock.setup(s => s.hasStoreData()).returns(() => hasStoreData);
        storesHubMock.setup(s => s.addChangedListenerToAllStores(It.isAny())).verifiable();
        storesHubMock.setup(s => s.removeChangedListenerFromAllStores(It.isAny())).verifiable();

        storesHubMock
            .setup(s => s.getAllStoreData())
            .returns(() => {
                return (
                    storeMocks && {
                        visualizationStoreData: storeMocks.visualizationStoreData,
                        tabStoreData: storeMocks.tabStoreData,
                        visualizationScanResultStoreData:
                            storeMocks.visualizationScanResultsStoreData,
                        featureFlagStoreData: storeMocks.featureFlagStoreData,
                        detailsViewStoreData: storeMocks.detailsViewStoreData,
                        assessmentStoreData: storeMocks.assessmentStoreData,
                        quickAssessStoreData: storeMocks.quickAssessStoreData,
                        pathSnippetStoreData: storeMocks.pathSnippetStoreData,
                        scopingPanelStateStoreData: storeMocks.scopingStoreData,
                        userConfigurationStoreData: storeMocks.userConfigurationStoreData,
                        unifiedScanResultStoreData: storeMocks.unifiedScanResultStoreData,
                        needsReviewScanResultStoreData: storeMocks.needsReviewScanResultStoreData,
                        needsReviewCardSelectionStoreData:
                            storeMocks.needsReviewCardSelectionStoreData,
                        cardSelectionStoreData: storeMocks.cardSelectionStoreData,
                        tabStopsViewStoreData: storeMocks.tabStopsViewStoreData,
                        dataTransferViewStoreData: storeMocks.dataTransferViewStoreData,
                    }
                );
            });
        return storesHubMock;
    }

    function setupGetDetailsRightPanelConfiguration(
        contentPanelType: DetailsViewRightContentPanelType,
        selectedPivot: DetailsViewPivotType,
        returnConfiguration: DetailsRightPanelConfiguration,
    ): void {
        const expected: GetDetailsRightPanelConfigurationProps = {
            selectedDetailsViewPivot: selectedPivot,
            detailsViewRightContentPanel: contentPanelType,
        };
        getDetailsRightPanelConfiguration
            .setup(gtrpc => gtrpc(It.isValue(expected)))
            .returns(() => returnConfiguration);
    }

    function setupGetSwitcherNavConfiguration(
        selectedPivot: DetailsViewPivotType,
        returnConfiguration: DetailsViewSwitcherNavConfiguration,
    ): void {
        const expected: GetDetailsSwitcherNavConfigurationProps = {
            selectedDetailsViewPivot: selectedPivot,
        };
        getDetailsSwitcherNavConfiguration
            .setup(gtrpc => gtrpc(It.isValue(expected)))
            .returns(() => returnConfiguration);
    }

    function setupActionMessageCreatorMock(
        mock: IMock<DetailsViewActionMessageCreator>,
        pivot: DetailsViewPivotType,
        timesCalled: number,
    ): void {
        mock.setup(acm => acm.detailsViewOpened(pivot)).verifiable(Times.exactly(timesCalled));
    }

    function getState(
        storeMocks: StoreMocks,
        viewType: VisualizationType,
        rightPanel: DetailsRightPanelConfiguration,
    ): DetailsViewContainerState {
        return {
            visualizationStoreData: storeMocks.visualizationStoreData,
            tabStoreData: storeMocks.tabStoreData,
            featureFlagStoreData: storeMocks.featureFlagStoreData,
            detailsViewStoreData: storeMocks.detailsViewStoreData,
            assessmentStoreData: storeMocks.assessmentStoreData,
            pathSnippetStoreData: storeMocks.pathSnippetStoreData,
            userConfigurationStoreData: storeMocks.userConfigurationStoreData,
            visualizationScanResultStoreData: storeMocks.visualizationScanResultsStoreData,
            scopingPanelStateStoreData: storeMocks.scopingSelectorsData,
            unifiedScanResultStoreData: storeMocks.unifiedScanResultStoreData,
            needsReviewScanResultStoreData: storeMocks.needsReviewScanResultStoreData,
            selectedDetailsView: viewType,
            selectedDetailsRightPanelConfiguration: rightPanel,
            cardSelectionStoreData: storeMocks.cardSelectionStoreData,
            needsReviewCardSelectionStoreData: storeMocks.needsReviewCardSelectionStoreData,
            permissionsStateStoreData: storeMocks.permissionsStateStoreData,
            tabStopsViewStoreData: storeMocks.tabStopsViewStoreData,
            cardsViewStoreData: storeMocks.cardsViewStoreData,
            quickAssessStoreData: storeMocks.quickAssessStoreData,
            dataTransferViewStoreData: storeMocks.dataTransferViewStoreData,
        };
    }
});
