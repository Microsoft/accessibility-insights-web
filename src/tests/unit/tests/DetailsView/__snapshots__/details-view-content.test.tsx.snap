// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` render renders normally 1`] = `
"<Fragment>
  <InteractiveHeader deps={{...}} selectedPivot={0} featureFlagStoreData={{...}} tabClosed={false} />
  <DetailsViewBody deps={{...}} tabStoreData={{...}} assessmentStoreData={{...}} pathSnippetStoreData={{...}} featureFlagStoreData={{...}} selectedTest={-1} detailsViewStoreData={{...}} visualizationStoreData={{...}} visualizationScanResultData={{...}} visualizationConfigurationFactory={[undefined]} assessmentsProvider={[undefined]} dropdownClickHandler={{...}} clickHandlerFactory={[undefined]} assessmentInstanceTableHandler={[undefined]} issuesSelection={[undefined]} issuesTableHandler={[undefined]} rightPanelConfiguration={{...}} switcherNavConfiguration={{...}} userConfigurationStoreData={{...}} cardsViewData={{...}} cardSelectionStoreData={{...}} scanIncompleteWarnings={[undefined]} scanMetadata={{...}} />
  <DetailsViewOverlay deps={{...}} previewFeatureFlagsHandler={[undefined]} scopingActionMessageCreator={[undefined]} inspectActionMessageCreator={[undefined]} detailsViewStoreData={{...}} scopingStoreData={{...}} featureFlagStoreData={{...}} userConfigurationStoreData={{...}} />
</Fragment>"
`;
